tools/workload-simulator.html

```diff
@@ -0,0 +1,1160 @@
+        if (deviceRequested) return;
+        deviceRequested = true;
+        let adapter = await navigator.gpu.requestAdapter();
+        featuresAndLimits.textContent = JSON.stringify(
+          {
+            name: adapter.name,
+            features: Array.from(adapter.features || []).slice().sort(),
+            limits: adapter.limits
+          }, 0, 2);
+        device = await adapter.requestDevice();
+        device.addEventListener('uncapturederror', (e)=>errorMessage.textContent = 'Uncaptured error: ' + e.error.message);
+        gpuPresent = webGPUCanvas.getContext('gpupresent');
+        swapChain = gpuPresent.configureSwapChain({device, format: swapChainFormat});
+        const bindGroupLayout = device.createBindGroupLayout({
+          entries: [
+            {
+              binding: 0,
+              // TODO: These shouldn't need to be visible to both stages. There's a bug in Dawn that
+              // incorrectly requires variables that are unused in one stage to be visible to all stages.
+              visibility: GPUShaderStage.VERTEX | GPUShaderStage.FRAGMENT,
+              buffer: { type: 'uniform', },
+            },
+            {
+              binding: 1,
+              visibility: GPUShaderStage.VERTEX | GPUShaderStage.FRAGMENT,
+              sampler: { type: 'filtering', },
+            },
+            {
+              binding: 2,
+              visibility: GPUShaderStage.VERTEX | GPUShaderStage.FRAGMENT,
+              texture: { sampleType: 'float', },
+            },
+          ],
+        });
+        const pipelineLayout = device.createPipelineLayout({
+          bindGroupLayouts: [bindGroupLayout],
+        });
+        if (multisampling.checked) {
+          multisampleRenderAttachment = device.createTexture({
+            size: { width: size, height: size },
+            sampleCount,
+            format: swapChainFormat,
+            usage: GPUTextureUsage.RENDER_ATTACHMENT,
+          })
+          if (!multisampleRenderAttachment)
+            errorMessage.textContent = 'Failed to allocate multisample render attachment.';
+        }
+        let shaderModule = device.createShaderModule({ code: `
+          let pos : array<vec2<f32>, 6> = array<vec2<f32>, 6>(
+              vec2<f32>(-1., 1.),
+              vec2<f32>(-1., -1.),
+              vec2<f32>(1., -1.),
+              vec2<f32>(-1., 1.),
+              vec2<f32>(1., -1.),
+              vec2<f32>(1., 1.));
+
+          [[block]] struct Uniforms {
+            xy: vec2<f32>;
+          };
+          [[binding(0), group(0)]] var<uniform> uniforms: Uniforms;
+          [[binding(1), group(0)]] var sampler1: sampler;
+          [[binding(2), group(0)]] var texture1: texture_2d<f32>;
+
+          struct Output {
+            [[builtin(position)]] Position : vec4<f32>;
+            [[location(0)]] uv : vec2<f32>;
+          };
+          
+          [[stage(vertex)]]
+          fn vs([[builtin(vertex_index)]] VertexIndex : i32) -> Output {
```

Comment:
has to be `u32` instead of `i32`

Reply:
Thanks. for some reason the error I was getting was a generic "can't parse WGSL" rather than a more detailed message, so I couldn't diagnose the problem.

Reply:
Naga produces good errors here. We are working on getting them to Gecko.